### PR TITLE
Support extension methods imported from different objects

### LINKED ISSUE #16920: Separate compilation breaks extension methods with the same name
## Compiler version

3.2.2 (and still in 3.3.0-RC2)

## Minimized code

Scastie: https://scastie.scala-lang.org/Ichoran/fgRflBYNRciNP6bBG9WtSw/14

```Scala
object One:
  extension (s: String)
    def wow: Unit = println(s)

object Two:
  extension (i: Int)
    def wow: Unit = println(i)

object Three:
  extension (s: String)
    def wow: Unit = println(s)
  extension (i: Int)
    def wow: Unit = println(i)

object Four:
  implicit class WowString(s: String):
    def wow: Unit = println(s)

object Five:
  implicit class WowInt(i: Int):
    def wow: Unit = println(i)

object Compiles:
  import Three._
  def test: Unit =
    5.wow
    "five".wow

object AlsoCompiles:
  import Four._
  import Five._
  def test: Unit =
    5.wow
    "five".wow

object Fails:
  import One._
  import Two._
  def test: Unit =
    5.wow
    "five".wow
```

## Output

```scala
value wow is not a member of Int.
An extension method was tried, but could not be fully constructed:

    Playground.Two.wow(5)    failed with

        Reference to wow is ambiguous,
        it is both imported by import Playground.One._
        and imported subsequently by import Playground.Two._

value wow is not a member of String.
An extension method was tried, but could not be fully constructed:

    Playground.Two.wow("five")    failed with

        Reference to wow is ambiguous,
        it is both imported by import Playground.One._
        and imported subsequently by import Playground.Two._
```

## Expectation

Extension methods work when compiled independently for different argument types (in different namespaces).

Thus, extension methods should be equally usable when compiled independently or jointly, and also are equally as usable as the implicit class extension mechanism.

## Actual Behavior

Independent compilation does not work with extension methods when the extension methods have the same name because the import ambiguity mechanism vetoes the successfully-resolved method call.  (Note that the error message tells you that it has actually fully successfully dispatched the method based on type.)

This prevents removal of the `implicit class` mechanism for extension methods, because `implicit class` does not have this problem.

(It seems that this problem should have been reported already by now, but I can't find it.  However, if this is not fixed, to be safe, nobody should use extension methods in any library, because it is impossible to predict when a collision might occur with some other library extending some other class. It's particularly acute with Java classes where you might want to define some obvious operation like `+`, but someone might want to define it on `Duration` and someone else might want to define it on `Path`.)

### PR DESCRIPTION
**Pull request by odersky**
Add a special case to name resolution so that when expanding an extension method from `e.m` to `m(e)` and `m` is imported by several imports on the same level, we try to typecheck under every such import and pick the successful alternative if it exists and is unambiguous.

Fixes #16920

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@mbovel it seems the benchmark server is down?

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
> it seems the benchmark server is down?

Fixed.

**Comment by odersky** [comment]
test performance please

**Comment by Ichoran** [comment]
This looks great!  I just wonder if it's necessary to add a test where an import is re-imported at a deeper level to make sure the search for alternatives is maintained.  Something like

```scala
object ReimportSucceeds:
  import One._
  def test: Unit =
    import Two._
    import One._
    "five".wow
```

The reason this is important is that people may want only some extensions in a broader context, but when they switch to a narrower context with more extensions, they may still wish to have the broader extensions available.

For instance, one might want `+` to add `Duration`s everywhere in a file, but to import a file-handling library that uses `+` on `Path` only in one method.  But one might want to do temporal math in that same method.

A bulk reimport is probably a reasonable balance between actual accidental collisions and requiring a bunch of extra boilerplate in order to get the desired functionality.  Does it already work this way by default?  Because I don't know how duplicate but more deeply-nested import statements are handled, I don't know whether this is actually a separate case that needs to be tested.

**Comment by odersky** [comment]
@ichoran Nested imports take precdence over outer ones, so the example you show should not be a problem.

**Comment by bishabosha** [comment]
doesnt this require a SIP?

**Review by bishabosha — CHANGES_REQUESTED** [review]
this should be under experimental import until SIP approval

**Comment by odersky** [comment]
I agree that this could be a SIP, but I won't have the time to submit and defend it. Would one of you want to take that over?

**Comment by sjrd** [comment]
I can write the SIP, yes.

**Comment by odersky** [comment]
For now this requires a language import
```scala
import language.experimental.relaxedExtensionImports`
```
We will drop that requirement once the SIP committee has approved the change.

**Comment by Ichoran** [comment]
@sjrd - I'm glad you're able to write the SIP!  If it would help to have more motivating examples, I have plenty!

**Review by bishabosha — CHANGES_REQUESTED** [review]
One of the other problems from [the Contributor's thread](https://contributors.scala-lang.org/t/change-shadowing-mechanism-of-extension-methods-for-on-par-implicit-class-behavior/5831) is that this example still does not work:
```scala
import language.experimental.relaxedExtensionImports
import scala.annotation.targetName

trait Foo[+T]

object Ops:
  extension (foo : Foo[Int])
    @targetName("bazInt")
    def baz : Unit = {}

import Ops.*

extension (foo : Foo[Double])
  @targetName("bazDouble")
  def baz : Unit = {}

@main def Test =
  val f = new Foo[Int] {}
  f.baz //error
```

In this case the `bazDouble` is always chosen because it is not imported, but again surely from the user's point of view it makes sense to also try the valid `Ops.bazInt` that was imported

**Comment by odersky** [comment]
> In this case the bazDouble is always chosen because it is not imported, but again surely from the user's point of view it makes sense to also try the valid Ops.bazInt that was imported. 

Yes, but that's just wishful thinking. And catering to that would simply produce a huge ball of mud. The current change is already highly problematic for it non-orthogonality. But the argument was that this is an important use case and there was no way to re-arrange things to make it work without the change. By comparison, the latest example is a fringe use case, not worth complicating the rules even further.

**Review by bishabosha — APPROVED** [review]
this is a great improvement over the status quo - and the local scope version always being picked is something that can be easily overcome by application developers

**Comment by sjrd** [comment]
SIP: https://github.com/scala/improvement-proposals/pull/60

File: docs/_docs/reference/contextual/extension-methods.md [hunk 1]
```diff
@@ -244,7 +244,18 @@ The precise rules for resolving a selection to an extension method are as follow
 Assume a selection `e.m[Ts]` where `m` is not a member of `e`, where the type arguments `[Ts]` are optional, and where `T` is the expected type.
 The following two rewritings are tried in order:
 
- 1. The selection is rewritten to `m[Ts](e)`.
+ 1. The selection is rewritten to `m[Ts](e)` and typechecked, using the following
+    slight modification of the name resolution rules:
+
+  - If `m` is imported by several imports which are all on the nesting level,
```
Comment (som-snytt) [inline docs/_docs/reference/contextual/extension-methods.md:250]: `all on the [same] nesting level`

Comment (odersky) [inline docs/_docs/reference/contextual/extension-methods.md:250]:
```suggestion
  - If `m` is imported by several imports which are all on the same nesting level,
```

**Comment by odersky** [comment]
Since SIP 54 just got accepted, I'll remove the experimental language import.

**Comment by sjrd** [comment]
It's been accepted to be shipped as experimental, at this stage. ;)

**Comment by odersky** [comment]
Ah indeed, it was labelled stage:design, not stage:implementation.

**Pull request merged**